File: src/main/java/io/github/emcw/map/entities/Town.java
package io.github.emcw.map.entities;

import com.google.gson.JsonObject;

import io.github.emcw.interfaces.ISerializable;
import lombok.Getter;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.awt.*;
import java.util.List;
import java.util.Objects;

import static io.github.emcw.utils.GsonUtil.*;

@SuppressWarnings("unused")
public class Town extends BaseEntity<Town> implements ISerializable {
    @Getter String mayor, nation;
    @Getter Integer area;
    @Getter Location location;
    @Getter List<Resident> residents;
    @Getter Flags flags;
    @Getter Color fill, outline;

    public Town(JsonObject obj) {
        super();
        init(obj);
    }

    public Town(@NotNull Capital capital) {
        super();

        setInfo(this, capital.getName());
        location = capital.getLocation();
    }

    void init(JsonObject obj) {
        setInfo(this, keyAsStr(obj, "name"));

        nation = keyAsStr(obj, "nation");
        mayor = keyAsStr(obj, "mayor");
        residents = Resident.fromArr(keyAsArr(obj, "residents"), "name");

        location = Location.of(obj);
        area = keyAsInt(obj, "area");
        flags = new Flags(obj);

        fill = getColour(keyAsStr(obj, "fill"));
        outline = getColour(keyAsStr(obj, "outline"));
    }

    public static class Flags {
        public final Boolean PVP, EXPLOSIONS, FIRE, CAPITAL, MOBS, PUBLIC;

        Flags(JsonObject obj) {
            PVP = keyAsBool(obj, "pvp");
            EXPLOSIONS = keyAsBool(obj, "explosions");
            FIRE = keyAsBool(obj, "fire");
            CAPITAL = keyAsBool(obj, "capital");
            MOBS = keyAsBool(obj, "mobs");
            PUBLIC = keyAsBool(obj, "public");
        }
    }

//    public Map<String, Resident> onlineResidents() {
//
//    }

    public boolean nationless() {
        return nation == null;
    }

    Color getColour(String hex) {
        return Color.decode(hex == null ? defaultColour() : hex);
    }

    String defaultColour() {
        return defaultColour(nation);
    }

    @Contract(pure = true)
    private static @NotNull String defaultColour(String nationName) {
        return Objects.equals(nationName, "No Nation") ? "#89C500" : "3FB4FF";
    }
}

File: src/main/kotlin/io/github/emcw/map/entities/Town.kt
package io.github.emcw.map.entities

import com.google.gson.JsonObject
import io.github.emcw.interfaces.IPlayerCollective
import io.github.emcw.interfaces.ISerializable
import io.github.emcw.utils.GsonUtil
import lombok.Getter
import org.jetbrains.annotations.Contract
import java.awt.Color

@Suppress("unused")
class Town : BaseEntity<Town?>, IPlayerCollective, ISerializable {
    @Getter
    var mayor: String? = null

    @Getter
    var nation: String? = null

    @Getter
    var area: Int? = null

    @Getter
    var location: Location? = null

    @Getter
    var residents: List<Resident?>? = null

    @Getter
    var flags: Flags? = null

    @Getter
    var fill: Color? = null

    @Getter
    var outline: Color? = null

    constructor(obj: JsonObject) : super() {
        init(obj)
    }

    constructor(capital: Capital?) : super() {
        setInfo(this, capital!!.name)
        location = capital.location
    }

    fun init(obj: JsonObject) {
        setInfo(this, GsonUtil.keyAsStr(obj, "name"))
        nation = GsonUtil.keyAsStr(obj, "nation")
        mayor = GsonUtil.keyAsStr(obj, "mayor")
        residents = Resident.Companion.fromArr(GsonUtil.keyAsArr(obj, "residents"), "name")
        location = Location.Companion.of(obj)
        area = GsonUtil.keyAsInt(obj, "area")
        flags = Flags(obj)
        fill = getColour(GsonUtil.keyAsStr(obj, "fill"))
        outline = getColour(GsonUtil.keyAsStr(obj, "outline"))
    }

    class Flags(obj: JsonObject?) {
        val PVP: Boolean
        val EXPLOSIONS: Boolean
        val FIRE: Boolean
        val CAPITAL: Boolean
        val MOBS: Boolean
        val PUBLIC: Boolean

        init {
            PVP = GsonUtil.keyAsBool(obj, "pvp")
            EXPLOSIONS = GsonUtil.keyAsBool(obj, "explosions")
            FIRE = GsonUtil.keyAsBool(obj, "fire")
            CAPITAL = GsonUtil.keyAsBool(obj, "capital")
            MOBS = GsonUtil.keyAsBool(obj, "mobs")
            PUBLIC = GsonUtil.keyAsBool(obj, "public")
        }
    }

    fun onlineResidents(): Map<String?, Resident?>? {
        return onlineResidents(residents!!, parent!!)
    }

    fun nationless(): Boolean {
        return nation == null
    }

    fun getColour(hex: String?): Color {
        return Color.decode(hex ?: defaultColour())
    }

    fun defaultColour(): String {
        return defaultColour(nation)
    }

    companion object {
        @Contract(pure = true)
        private fun defaultColour(nationName: String?): String {
            return if (nationName == "No Nation") "#89C500" else "3FB4FF"
        }
    }
}